### ISSUE TITLE
Discussion of "Traversing the Tree" and Overlapping Hierarchies

### ISSUE THREAD

**Issue by ebeshero** [ebeshero/DHClass-Hub#325]
### The reading:

Read Gabrielle Kirilloff, “[&lt;Traversing_the_Tree/&gt;](http://gk.obdurodon.org/seminar/seminarPaper.xhtml)”, the short article, “[Frankenstein novel analyzed](http://www.idsnews.com/article/2015/04/frankenstein-novel-analyzed)” and scroll through Wendell Piez's [conference talk and images for the Balisage Markup Conference 2014](https://www.balisage.net/Proceedings/vol13/html/Piez01/BalisageVol13-Piez01.html). If you like, you can take a close look at [his LMNL code of Frankenstein](https://github.com/wendellpiez/Luminescent/blob/master/lmnl/frankenstein-as-published.lmnl) on GitHub. 
_Special note of interest:_ Gabi Kirilloff was a student in [a digital humanities course at Pitt like the one you are taking](http://dh.obdurodon.org), and she originally wrote &lt;Traversing_the_Tree/&gt; for a seminar paper assignment in another class. 
### The discussion prompts:
- What perspective does Kirilloff provide on the kinds of XML markup we are learning, the history and context of hierarchical markup?
--What problems does hierarchical markup pose for encoding documents? What ideas do Kirilloff or Piez present for how to deal with these in code, and how effective or problematic might these be? Is it possible to write XML to "get around" the problems raised in these pieces? What's lost (or gained) in making XML's hierarchical structure deal with overlap?
- Consider the examples of overlapping hierarchies that Kirilloff and Piez present to us: Which of these did you find especially interesting? Are there good ways to "model" overlapping hierarchies with code? 

The discussion is worth points to you as a homework exercise. For full credit, your posts should make specific reference to passages in Kirilloff's essay, and reflect on those passages. **Note:** You (as an individual) do not have to respond to every one of the discussion prompts, but our two classes as a whole should cover them all. A good strategy would be to try to respond in some way to at least two of the prompts in the list above. Raising questions is encouraged, and so is responding to each other, but responding should do more than simply say, "yes, I agree." A good response will add something new to the conversation, or help promote more discussion. You should make **at least two substantial posts** to fully contribute to the discussion.

*As you're drafting your comments, see if you can apply "Markdown" formatting if you'd like to use bold or italics or make a list, form a link, add an image, etc. Follow the link to ["Styling with Markdown is supported"](https://guides.github.com/features/mastering-markdown/) (which you can always find at the bottom left of an Issue write screen) for an orientation to Github's markdown.*

**Comment by quantum-satire** [comment]
**Addressing: What perspective does Kirilloff provide on the kinds of XML markup we are learning, the history and context of hierarchical markup? What problems does hierarchical markup pose for encoding documents? Is it possible to write XML to "get around" the problems raised in these pieces?**
— — — 
*Traversing the Tree* explores a few approaches to the types of markup languages used for digital humanities; specifically XML, SGML, and OHCO. XML is a derivative of SGML, which was developed in the 80’s for the purposes of encoding data and exchanging readable texts between computing machines, and was created with the intention of making documents accessible through the digital storage for industrial and governmental use. The language however would be based upon the genre of the text in question; it was never intended for use in the digital humanities, only for the practical applications previously listed, and the focus was placed heavily on the ease of use and accessibility of the overall structure. The issues that arose with XML and SGML were considered unavoidable:

SGML | XML
------------ | -------------
HTML works better for encoding texts | No predefined tag sets; the coder is tasked with presenting the tags and applying them, possibly leading to inconsistencies
Originally intended to represent a single hierarchy in a document | Computers process XML as binary code and do not comprehend the text as humans do
Hierarchies are dependent on genre | Shares the same logic as SGML, which was not intended for use in digital humanities

There are also issues that came to be for both languages:
* Hierarchy is dependent on the interpretation of the text, and is impossible to have one single standard form for all circumstances
* Overlapping hierarchies occur frequently, moreso in extensive paragraphs and metaphors (highly applicable to the humanities)

As for writing XML to skirt the problems, Kirilloff presents a way to address the problem of overlapping hierarchies including the usage of empty elements, but the solution falls short of actually acknowledging the existence of the overlapping hierarchies. She writes on how others have attempted to address the issue similarly, but that none have been able to fix the form itself, therefore answering the posed question.

**Comment by kes213** [comment]
_**Addressing:**_ **Consider the examples of overlapping hierarchies that Kirilloff and Piez present to us: Which of these did you find especially interesting? Are there good ways to "model" overlapping hierarchies with code?**

	Of the examples of overlapping hierarchies that Kirilloff provides, I found the overlap that occurs when a paragraph extends over multiple pages. This example particularly caught my attention due to my writing background. I’m a Creative and Professional Writing major, and I’m used to being long-winded. When I’m writing a story that I’m extremely passionate about -- fictional, journalistic, or poetic -- I often find that I have a lot to say. When I’m writing a news story, trying to fit all of the information in the piece that I possibly can, it’s easy to forget to split my thoughts into paragraphs. I just have so many thoughts running through my head that I can’t pause for a second to tab down to a new paragraph -- I could lose my train of thought, and then where would I be? 
	Even now, as I type this, I find myself rushing to make sure I don’t forget what I want to say. 
	As a general rule in novel publication, a paragraph should be shorter than one page. If it’s longer, it needs to be split somewhere to give the reader a break, a chance to pause in the writing to process information and prepare for another point, description, example, etc. In journalistic writing, paragraphs should be even shorter -- sometimes no longer than three sentences each, sometimes just long enough for a single quote. That’s even more of a struggle for me. 
	When Kirilloff suggests that overlapping hierarchies include a paragraph that stretches over a few pages, I started thinking about the historical letter XML exercise that we did as a class. My strategy was to divide the letter first into paragraphs and second into sentences. I think this is an example of overlapping hierarchies in itself, and I wondered if organizing a letter or a long document by paragraph would make sense if the paragraphs extended over pages. Should I identify what page the paragraph is on, and therefore create an overlapping hierarchy? Or should I mark the documents more simply and stick with just paragraph and sentence? When would it be helpful to have a page number? 
	“Research papers” was my answer. I think we’ve all been there, whether you’re a writing major or not. You find an amazing quote that you want to put in your research paper, but you forgot to highlight it in your source, and now you can’t find it. A quick search of attributes would find the page number for you in a snap, and now you don’t have to read a 100-page source all over again to find the quote that’s probably on page 99 if you have luck like I do. Overlapping hierarchies can certainly be helpful, even with things like historical documents or other long works. 

	As a writer, I ultimately look at a piece of literature and focus on _content_ rather than _form_, except in rare cases in poetry when a poem’s form carries significant meaning from the poem. I think that coding a text based on content -- grouping main ideas together with examples, marking each new reason or idea so that they are distinct and easily noted -- is a more effective way to keep the meaning of the text alive within a computer code. I would support the ontological model that Kirilloff describes for that reason -- text is more than a heading and a body and a conclusion. Without the actual words within the text, there’s no reason for the rest of that format to exist. If the sentences don’t express ideas, there’s really no text to code, because if the text doesn’t have meaning, why bother storing it for future use? 

In short, I’m an organization freak -- I color-code _everything_, I alphabetize my DVD collection. I also just love words, so I think organization should revolve around them.

**Comment by kes213** [comment]
@quantum-satire, I'd like to add to your comment! 

Piez provides a more detailed examination of the overlapping hierarchies within poem lines. Piez points out that a poem can be divided by stanzas and lines but also by grammatical structures like sentences and phrases. Kirilloff touches base on this in “Traversing the Tree,” as well, but Piez provides more examples. While I think in some ways, SGML is more “logical” in that it is more consistent with tag names, XML fits better for poetry. There isn’t just _one_ form of poetry -- a poem can be as long or as short as the poet wants it to be; it can rhyme, or it can be free verse. The lines can all be of equal meter and measure, or maybe line 1 is three syllables and line 10 is 8 syllables because the poem is about an ocean and the poet wants the lines to look like waves washing up on the shore as the reader looks at the page. Piez mentions a few types of poems that keep to a strict structure, such as sonnets, which rely heavily on form for organization. For many other poems, though, using XML is perfect because it lets the coder label each tag in any way he or she sees fit; not every person reads a poem and gets the same message. XML has no “predefined” tags, like you said, and for that reason, XML is more applicable to multiple writing genres.

**Comment by ebeshero** [comment]
@kes213 @quantum-satire Thus far, your posts are reminding me that markup and coding is after all a *form* of writing! We use the syntax of markup to organize text for ourselves...and perhaps might apply two different markup approaches to show us different features of the same document?

**Comment by tal80** [comment]
**Addressing:  What perspective does Kirilloff provide on the kinds of XML markup we are learning, the history and context of hierarchical markup?**

While reading Traversing the tree by Kirilloff, we encounter various types of the markup methods used within the Digital Humanities field. Those encounter include: SGML, XML, OHCO. SGML (Standard Generalized Markup Language) was created before XML. SGML was intended for the facilitation of the sharing and storage of large-project documents in law, government, and industry. The context of hierarchical markup differed between SGML and XML differs from one another. I agree with Kes213 in that XML make a better fit for poetry. The reason being, freely creating tags.  SGML is more logical and defined in the way that it uses predetermined tags. In short, use XML if you wish to use personally defined tags, but beware you could be inconsistent in your use of them. Using XML will allow you to encode different genres of whatever medium you are using which in my case is Primary Historical sources. Use SGML for a more defined model of mark up, as mentioned above, government projects, law, and industry. I'm stuck as to which would be better in my case. I'm used to primary historical sources. My guess would be SGML. What do all of you think?

**Comment by pab124** [comment]
Addressing: What perspective does Kirilloff provide on the kinds of XML markup we are learning, the history and context of hierarchical markup?
Being a Lingusitics major, one of the first classes that I took related to my major was Accelerated Syntactic Theory. After we got past the very basic 'defining parts of speech' and all that fun stuff, we began to discuss syntactic processes and the structure that language takes. One of the more important topics and methods discussed was the Syntax Trees that we made so many of. Reading the Traversing the Tree text reminded me of similar problems that occurred in describing some syntactic patterns. There were many instances where there should be a clear hierarchy in the constituents of a sentence, but the tools (syntactic rules) we were given were not quite enough to effectively explain or map out the characteristics of the sentence. Similarly, Kirilloff makes note of several instances where the hierarchical structure of the text presents difficulty in the coding of the data. Since the syntax class was the first core class that I was exposed to, naturally I sort of fell into that train of thought. Eventually, there came ways that we were able to apply the rules, or variants of them, that would yield a correct, though often times clunky and cumbersome, solution. Would this be a similar situation to the attempt to use XML to 'get around' some of the hierarchical problems? Though the solutions may be correct and acceptable, the methods used may detract from the elegance that the theory seeks.

**Comment by ghbondar** [comment]
@pab124 Interesting... can you explain, to us non-linguists, what is a syntax tree? I was just observing to @ebeshero how in English, the grammar of a sentence often follows an OHCO model.

**Comment by ebeshero** [comment]
@tal80 SGML was an early form before XML existed, and it permitted multiple trees in the same document--it *could* mean you'd have multiple root elements. XML replaced it, as I understand, because it was easier to process a document made of a single tree structure. That said, XML then has the problem Kirilloff is identifying with overlapping hierarchies...

**Comment by pab124** [comment]
@ghbondar basically, a syntax tree is a tool used to produce a diagram of a parsed sentence. It breaks down each element of the sentence and places it into its own node. after completing a tree, there are terminal nodes (named for the part of speech they contain) that hold each word in a sentence. an example of this would be the Noun Phrase (NP) The Red Dog. Here, there are 3 elements that are nested within. The (determiner) Red (Adjective Phrase) Dog (Noun). Likewise, the ADJP is broken down further to display the ADJ that is in it. This NP may be contained within a Verb Phrase if it's the object of a verb, which in turn would be part of a Tense Phrase, which in turn would be part of a Complementary Phrase, which is essentially the Root Element of a syntax tree. Many elements are nested within other elements, which in turn are nested within a final, all covering Head Node, very similar to the tagging and parsing we've done this far in XML. However, certain problems with the trees can arise in ambiguity (as in "The British left waffles on Falklands") and there are some alternative methods to account for this, just like there are some methods for producing correct code when hierarchical situations arise in coding. Hopefully this made sense, it's rather difficult to explain in a paragraph.

**Comment by ttb11** [comment]
Consider the examples of overlapping hierarchies that Kirilloff and Piez present to us: Which of these did you find especially interesting? Are there good ways to "model" overlapping hierarchies with code?

In XML, the example given in the reading, "<line><word>eng-</line>ineers</word>", is incorrect. It isn't well formed as the elements aren't neatly stacked. This problem in XML form is similar to a lexical gap in linguistics. For example, there's no real word for "shallow" in French. Instead, you must say "peu profond" which translates directly to "little deep." Similarly, in XML there is no standout answer for representing overlapping hierarchies.

Also, in the paper the question “What is text?” is raised. This seems like such an easy question until you start to think about it. Similarly in morphology we ask, "What is a word?" Both of these seem like they have easy answers but are "dog," dogs," "dog's," and "dogs'" all separate words? Or are they iterations of one word? The abstract idea of text as well as the ambiguity of the word "text" also raises questions.

Finally in the conclusion, the author talked about CONCUR, an alternative to OHCO. This made me realize just how much coding is language. In the "lexical gap" found in XML and SGML, we can find an historically used "word," CONCUR, that used in SGML that addressed the problem. Similarly in english, historically we had words that corresponded to plural-you "you," and singula-you "thou," the latter of which has fallen out of common use. While CONCUR didn't disappear, it is a clunkier, less used version of XML, (like though is an awkward, less used version of you.)

**Comment by ttb11** [comment]
@pab124 A lot of code also reminds me of linguistics. Although I haven't yet taken syntax, I did do syntax trees in Intro. That was actually the example I was going to use for my response before I saw it in yours. To answer the question raised in your response, I think that getting a clunky outcome is better than a wrong one. If overlap is not possible in XML it is better to be able to awkwardly represent it than to be incapable of representing it at all. This is similar to the example I used in french, it's better to be able to say "peu profond" for shallow instead of not having a word for shallow at all.

**Comment by quantum-satire** [comment]
After thinking about the issues that need to be worked around while using these languages, I wonder if it wouldn't be better to start another kind of code entirely with the sole purpose of encoding texts for the Digital Humanities? Because SGML was created to encode industrial and governmental texts, it had a singular purpose that worked as a catch-all for those types of documents because they all had a similar structure or style; however, because the Humanities have so many different types of texts (poems, novels, etc.) I wonder if a new brand of code could be developed with interchangeable pieces to fit each circumstance--you could have the same type of elements that are able to be pieced together to form all types of structures to fit the piece you are working with, (and to be able to set defined tags to read the entire document and mark words that fit those parameters, but I am not certain if XML is capable of doing this).

As @ttb11 wrote
>To answer the question raised in your response, I think that getting a clunky outcome is better than a wrong one. If overlap is not possible in XML it is better to be able to awkwardly represent it than to be incapable of representing it at all. This is similar to the example I used in french, it's better to be able to say "peu profond" for shallow instead of not having a word for shallow at all.

I think that instead of attempting to use the language awkwardly and "deal" with these issues, we can take what we've learned so far about text markup for the Humanities and turn it into something better suited for our use. I'm probably not the only person who has said this before and I realise how difficult this may be, but I'm certain there is a way to fix this without simply settling for the clunky (and temporary) way!

**Comment by BMR59** [comment]
Addressing: Consider the examples of overlapping hierarchies that Kirilloff and Piez present to us: Which of these did you find especially interesting? Are there good ways to "model" overlapping hierarchies with code?

I found this section particularly interesting because I struggled with it for my XML 03 homework - I attempted to code a passage from The Aeneid but wound up lost on how to remedy overlapping hierarchies between the lines of the epic. From what I gathered, empty elements are a solution to the issue, but also create their own issue by not acknowledging the overlapping hierarchy. Being completely new to coding, it's a lot to take it but I'm enjoying learning the processes and formulas of XML so far. The other option appears to be using OHCO, though it is method of coding that has fallen from common use and is more technical.

**Comment by djbpitt** [comment]
Voice of History, here, since I used SGML in production back in the 1980s, before there was XML. @kes213 writes that "I think in some ways, SGML is more 'logical' in that it is more consistent with tag names", and @tal80 adds that "SGML is more logical and defined in the way that it uses predetermined tags". I'm curious about what it is in the reading that suggests that SGML "generic identifiers" (GIs, element or tag names) might be more consistent than those in XML, or that SGML uses predetermined GIs. SGML and XML follow the same basic rules, neither uses any predetermined GIs, and both allow users to create their own GIs. This is a difference from HTML, which uses only a fixed inventory of predefined GIs.

One important difference between XML and SGML is that XML markup is case-sensitive and SGML is case-insensitive, and in that sense it would seem as if it is XML that is more consistent. On the other hand, that consistency creates an opportunity for confusion or user error, since XML case-sensitivity means that you can create *different* elements called, for example, `<Title>` and `<title>`, while in SGML those would refer to the same element type.

**Comment by flowerbee1234** [comment]
Consider the examples of overlapping hierarchies that Kirilloff and Piez present to us: Which of these did you find especially interesting? Are there good ways to "model" overlapping hierarchies with code?

One of the things I really enjoyed in Kirilloff's article is his emphasis on interpretation. He claims that the scholar's incorporation of the text can change its meaning, and I definitely agree. To begin, Kirilloff expands upon the idea that digital humanities adds to humanities itself. He gives some historical background of, say, XML, which dates back to the '80's and '90's. XML grew out of SGML to be used in "law, government, and industry." Now, coders can use XML for operations such as digital humanities, to work with literary pieces. Like never before, the humanitarian can take these art pieces and dissect them in newer, more technologically-savvy ways. 

Now I should touch on the overlapping concept. By using the tree-method, one can avoid two elements from overlapping, having a parent, children, siblings, and so forth. This can be commonly found when coding poems, since the formatting can be both complicated and sophisticated. If the writer formats his piece for a structural purpose, then coding it may cause some problems. In fact, as previously mentioned, it can change the overall meaning of the writing. I found this to be interesting. 

There are several solutions, as per Kirilloff, to avoid this overlap. Primarily, the coder can change the syntax and use a slash before the closing bracket: <tag/>. This, then, "tricks the system" and is simply an alternative method. One could also use OHCO to avoid these altogether. I think I will, to avoid overlapping, use the tree method and simply branch things out from each other. This way, things can be organized and give a "clean" look. I believe nothing is worse the confusion caused by overlap and giving Oxygen a heart attack, lol.

**Comment by flowerbee1234** [comment]
@kes213 I like your view of content and form. I believe, as writers, we can mess with the content and form and emphasize it through coding. Of course, like you said, form seems to matter more in poetry--but form is also important in prose. For example, the lengths of paragraphs, the amount of dialogue, etc. These things are all important when looking at a piece.

**Comment by Jamielynn92** [comment]
Consider the examples of overlapping hierarchies that Kirilloff and Piez present to us: Which of these did you find especially interesting? Are there good ways to "model" overlapping hierarchies with code?
   
     As someone who is completely new to the world of XML and overlapping hierarchies and SGML, if I had to choose one to use, it would XML. Although I'm not sure if XML is able to overlap, but the way it looks using SGML, looks a lil bit confusing to me. But is it interesting on their two point of views. Even though they both agreed that there is no single "standard" hierarchy. I'm not sure what that really means, but for me(whos never worked with either before) thinks that it would be a little bit easier with XML. But if I got the chance to learn the other way, I'm not sure if it would make it easier or harder to 'communicate' with the deep computering world of coding.

The relationship between OHCO and XML would make a good way to show overlapping hierarchies. it may be a bit tricky, especially since I'm not totally sure on how to do that, but I'm sure its exciting when it happens. I feel like it would require many inside tags with each element created. wouldn't it become confusing.

**Comment by Jamielynn92** [comment]
What perspective does Kirilloff provide on the kinds of XML markup we are learning, the history and context of hierarchical markup?

The perspective made upon XML markup is that it is a machine-readable way to code all different kinds of documents. You can add different <tags> within a document using XML and it might look normal. XML was suppose to have different goals than that of SGML. It was developed in the late 80s, and is a way of another language. But soon after XML was the chosen choice. It may have made it a little bit 'unavoidable'. But they share formal logic.
The different types of ways that scholars describe texts are SGML, XML and TEI. They include their own set of rules. Its like a language all its own and its an ordered hierarchies of content objects(OHCO). Tree structures are what the hierarchies form into. With the individual components forming relationships like a tree. XML rules involve forming and maintaining hierarchy structure. If it can't follow the rules, its nothing. No over lapping can occur. Must be nested within the markup order. The tree works very well.

**Comment by ebeshero** [comment]
Thanks, everyone, for your thoughtful posts! It seems several of you are curious about SGML--and I've only seen it from a distance but haven't worked with it. (Thanks, @djbpitt , for visiting and commenting on the interesting matter of its tag name conventions!) By the way, @djbpitt is a good friend of our course, and indeed taught me and @ghbondar our first XML code some years ago, and he is indeed, the author of the Pittsburgh DH course material at http://dh.obdurodon.org , a close relative of our course at http://dh.newtfire.org . 

As of now, we don't really have much of a *practical* choice now between these technologies, if we want to be concentrating on doing markup to build digital projects, rather than reinventing the language and tree structure of XML. That is because there are *communities of practice* now that have put their energies into developing support technology for XML as you have been learning to write it. An XML document gives us a single tree to process, with its single root element, and while it might be just fine to represent, er, a shopping list or a recipe, it's definitely a limited, problematic way to try to represent complex writings like poetry, where much of the joy of reading may come from the friction or irony when we detect an overlap in distinct hierarchies--for example, what we experience when sentences end in the middle of a line, or when rhyming sounds at the ends of lines are sometimes the action verbs in a longer sentence weaving through the lines. How could we use XML, with its one tree structure to represent such complexity? 

Well, I have to admit, in the several XML projects I've built in the past few years, I have probably enjoyed "tricking" the tree structure more than I should. I use self-closing "milestone-style" elements all the time with attributes that indicate the start and end of some interesting phenomena that is literally tangled around a formal structure. I'm fascinated by that, really, and I do tend to write the tree structures show they show the most regular form I can see in a document. But it's important to realize our code is kind of our interpretation, something we apply *overtop* or *around* a document, and we can always change it, pull it out, write it *differently*. 

I think one answer to the problem posed by OHCO (the Ordered Hierarchy of Content Objects) is simply to *write some alternative code*: build another tree, make multiple XML codings to help show different co-existing orders. We might bind these documents together using attributes to show us where the self-closing elements in one correspond to the open text-containing element structure of the other. That might be useful, depending on what you want from a markup project. Usually, though, our projects just have one tree that we design for a clear purpose--and we find ways to work with (compromise with!) the code-base we have, so that the whole we come up with is somehow greater than the sum of its parts. By that I mean, *usually* the projects (the "digital castles" or "cathedrals" we build from angle-bracket element "bricks") show something more complex than the code-base by itself seems to support.
